### PR TITLE
Adds Optimizer support and local epoch count for federated simulations.

### PR DESCRIPTION
**Pull request by derikclive**
PR adds support for the Adam, AdaGrad and SGD with momentum for the federated simulations experiments. Also, support for modifying the number of local epochs per federated round has been added. Util.py file has been split into 2 files one for data utilities and the other for model utilities.

### CODE REVIEW

File: Federated Model (seq-to-seq)/model_utils.py [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+    if prev_tag != 'O' and prev_tag != '.' and prev_tag_type != tag_type:
+        chunk_end = True
+    return chunk_end
+
+
+def split_tag_type(tag):
+    """Extracts and returns the tag and tag type.
+    """
+
+    split_tag = tag.split('-')
```
Comment (joezhoujinjing) [inline Federated Model (seq-to-seq)/model_utils.py:166]: maybe use re.match?

File: Federated Model (seq-to-seq)/model_utils.py [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+    if len(split_tag) == 1:
+        tag = split_tag[0]
+        tag_type = ""
+    else:
+        tag = split_tag[0]
+        tag_type = split_tag[1]
+    return tag, tag_type
+
+
+def compute_f1(correct_slots, pred_slots):
```
Comment (joezhoujinjing) [inline Federated Model (seq-to-seq)/model_utils.py:224]: This function seems to be a bit long and hard to follow, could you break it into smaller pieces?

File: Federated Model (seq-to-seq)/model_utils.py [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+    semantic_acc = (intent_pred == intent_real)
+
+    for idx, (s_pred, s_real) in enumerate(zip(slot_pred, slot_real)):
+
+        for i in range(len(s_real)):
+            if s_real[i] == 0:
+                break
+
+            if s_pred[i] != s_real[i]:
+                semantic_acc[idx] = False
```
Comment (joezhoujinjing) [inline Federated Model (seq-to-seq)/model_utils.py:326]: also break here?

File: Federated Model (seq-to-seq)/train_model.py [hunk 4]
```diff
@@ -10,23 +10,39 @@
...
+np.random.seed(123)
+
+
+def _optimizer_canonical_name(optimizer_cls):
+    """Return a short, canonical name for an optimizer for us in flags."""
+    return optimizer_cls.__name__.lower()
+
+
+# List of optimizers currently supported.
+_SUPPORTED_OPTIMIZERS = {
```
Comment (joezhoujinjing) [inline Federated Model (seq-to-seq)/train_model.py:39]: It might be clearer if you just hardcode the key with "sgd", "adagrad", "adam"?

File: Federated Model (seq-to-seq)/train_model.py [hunk 5]
```diff
@@ -249,53 +323,78 @@ def create_keras_model(arg,
...
-        from_logits=True, reduction='none')
-    mask = tf.math.logical_not(tf.math.equal(y_true, 0))
-    slot_losses = loss_objective(y_true, y_pred)
+    server_opt_cls = _SUPPORTED_OPTIMIZERS.get(arg.server_optimizer)
+    client_opt_cls = _SUPPORTED_OPTIMIZERS.get(arg.client_optimizer)
+
+    if arg.server_optimizer == 'sgd':
+        server_opt = lambda: server_opt_cls(learning_rate=arg.server_lr,
+                                            momentum=arg.momentum)
+    else:
```
Comment (joezhoujinjing) [inline Federated Model (seq-to-seq)/train_model.py]: better to be more explicit, i.e elif arg.server_optimizer in ['adam', 'adagrad']:
And add a not support error in the else clause.

**Review by joezhoujinjing — APPROVED** [review]

File: Federated Model (seq-to-seq)/model_utils.py (hunk 1 above)
Comment (derikclive) [reply Federated Model (seq-to-seq)/model_utils.py:166]: I think split works better here since we need to get the slot name and the slot type. I'm not sure if using re.match would be better.
